iml-config/src/nginx.rs

```diff
@@ -0,0 +1,133 @@
+    "TIMER_PROXY_PASS",
+    "INCLUDES",
+];
+
+fn replace_template_variables(contents: &str, template_vars: &[&str]) -> String {
+    let mut updated_contents: String = contents.to_string();
+    for var in template_vars {
+        updated_contents = updated_contents.replace(
+            // find a better way to do this
+            &format!("{{{{{}}}}}", var),
+            &env::var(var).unwrap_or_else(|_| panic!("{} variable not set", var)),
+        );
+    }
+
+    updated_contents
+}
+
+pub async fn nginx_cli(command: NginxCommand) -> Result<(), ImlConfigError> {
+    match command {
+        NginxCommand::GenerateConfig {
+            template_path,
+            output_path,
+        } => {
+            let nginx_template_bytes = fs::read(template_path).await?;
+            let nginx_template = String::from_utf8(nginx_template_bytes)?;
+
+            let config = replace_template_variables(&nginx_template, TEMPLATE_VARS);
+
+            if let Some(path) = output_path {
+                fs::write(path, config).await?;
+            } else {
+                let term = Term::stdout();
+
+                tracing::debug!("Nginx Config: {}", config);
+
+                term.write_line(&config).unwrap();
+            }
+        }
+    };
+
+    Ok(())
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use insta::*;
+
+    fn clear_env_vars(vars: &[&str]) {
+        for key in vars {
+            env::remove_var(key);
+        }
+    }
+
+    #[test]
+    fn test_replace_template_variables() {
+        let template: &[u8] = include_bytes!("../../chroma-manager.conf.template");
+
+        clear_env_vars(TEMPLATE_VARS);
+
+        env::set_var("REPO_PATH", "/var/lib/chroma/repo");
+        env::set_var("HTTP_FRONTEND_PORT", "80");
+        env::set_var("HTTPS_FRONTEND_PORT", "443");
+        env::set_var("HTTP_AGENT_PROXY_PASS", "http://127.0.0.1:8002");
+        env::set_var("HTTP_AGENT2_PROXY_PASS", "http://127.0.0.1:8003");
+        env::set_var("HTTP_API_PROXY_PASS", "http://127.0.0.1:8001");
+        env::set_var("IML_API_PROXY_PASS", "http://127.0.0.1:8004");
+        env::set_var("WARP_DRIVE_PROXY_PASS", "http://127.0.0.1:8890");
+        env::set_var("MAILBOX_PATH", "/var/spool/iml/mailbox");
+        env::set_var("MAILBOX_PROXY_PASS", "http://127.0.0.1:8891");
+        env::set_var("SSL_PATH", "/var/lib/chroma");
+        env::set_var("DEVICE_AGGREGATOR_PORT", "8008");
+        env::set_var("DEVICE_AGGREGATOR_PROXY_PASS", "http://127.0.0.1:8008");
+        env::set_var(
+            "UPDATE_HANDLER_PROXY_PASS",
+            "http://unix:/var/run/iml-update-handler.sock",
+        );
+        env::set_var("GRAFANA_PORT", "3000");
+        env::set_var("GRAFANA_PROXY_PASS", "http://127.0.0.1:3000");
+        env::set_var("INFLUXDB_PROXY_PASS", "http://127.0.0.1:8086");
+        env::set_var("TIMER_PROXY_PASS", "http://127.0.0.1:8892");
+        env::set_var("INCLUDES", "");
+
+        let config = replace_template_variables(
```

Comment:
The test is heavily relying on the environment, could be possible to abstract defining the var from substitution. E. g. `replace_template_variables(template: &str, vars: &HashMap<...>)`.

Reply:
Technically, just pass all the variables to template.

Reply:
Each var is repeated in at least three places (template, test, TEMPLATE_VARS), this can easily go sideways, I think.

Reply:
As of now the idea is that this will be used by a service that has a loaded environment file so I think testing with the environment makes sense.

Reply:
I mean that if we need to change vars (add, remove, rename), we will have to do in multiple places and a missing var will not be caught in production (until nginx fails), nor in tests. A kind of speculation, but if template included some raw string (not a var ref), the test would pass anyway :) To be more reliable there better be at least two tests with different values of variables.

Reply:
In addition, tests are run within the same processes in multiple threads. Setting the env here means other tests can be effected.